systemtest/cmd/apmbench/main.go

```diff
@@ -94,11 +94,19 @@ func BenchmarkAgentRuby(b *testing.B, l *rate.Limiter) {
 }
 
 func benchmarkAgent(b *testing.B, l *rate.Limiter, expr string) {
+	parentHandler := benchtest.NewEventHandler(b, expr, l)
 	b.RunParallel(func(pb *testing.PB) {
-		h := benchtest.NewEventHandler(b, expr, l)
+		h := benchtest.NewPooledEventHandler(expr, func() any {
+			h, err := parentHandler.Clone()
+			if err != nil {
+				b.Fatal(err)
+			}
+			return h
+		})
 		for pb.Next() {
 			h.SendBatches(context.Background())
 		}
+		benchtest.ReleaseEventHandler(expr, h)
 	})
```

Comment:
This seems awkward, since it's adding to a global pool (i.e. updating the pool map) - but the handler is parameterised with a rate limit and that's not part of the map key. Is the map necessary?

Would it be possible to simplify to something like this?

```go
pool := sync.Pool{New: func() any { return benchtest.NewEventHandler(b, expr, l) }}
b.RunParallel(func(pb *testing.PB) {
    h := pool.Get().(*eventhandler.Handler)
    defer pool.Put(h)
    defer h.Reset()
    // use h
})
```

Reply:
I was trying to not have to change the signature of the current `BenchmarkFunc`. I decided to use the map, since currently `benchtest.NewEventHandler` filters the files using a the provided glob expression. If we're indiscriminately storing the event handlers into a single pool, then the wrong handlers will be re-used.

If we implement as you suggest, then the pool won't survive after the benchmark function has exited, since it's local to that function, so the pool won't be very helpful.

Another option would be to have `NewEventHandler` scan all the files and store the compressed batches in a `map[string][]batch`. That would allow us to use a single pool and re-initialize the event handler and reset it just before putting it back into pool.

Reply:
> If we implement as you suggest, then the pool won't survive after the benchmark function has exited, since it's local to that function, so the pool won't be very helpful.

Ah - I was thinking that they were already local to the function, but I see now.

> Another option would be to have NewEventHandler scan all the files and store the compressed batches in a map[string][]batch. That would allow us to use a single pool and re-initialize the event handler and reset it just before putting it back into pool.

This sounds sensible to me.